stablehlo/dialect/StablehloTypeInference.h

```diff
@@ -0,0 +1,162 @@
+/* Copyright 2022 The StableHLO Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef STABLEHLO_DIALECT_STABLEHLO_TYPE_INFERENCE_H
+#define STABLEHLO_DIALECT_STABLEHLO_TYPE_INFERENCE_H
+
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/StringRef.h"
+#include "mlir/Dialect/Quant/QuantTypes.h"
+#include "mlir/Dialect/Shape/IR/Shape.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/IR/DialectImplementation.h"
+#include "mlir/IR/Location.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/TensorEncoding.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/IR/Types.h"
+#include "mlir/Interfaces/InferTypeOpInterface.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+#include "mlir/Support/LogicalResult.h"
+#include "stablehlo/dialect/Base.h"
+
+#ifndef ENUMS_AND_ATTRS
+#define ENUMS_AND_ATTRS
+#include "stablehlo/dialect/StablehloEnums.h.inc"
+#define GET_ATTRDEF_CLASSES
+#include "stablehlo/dialect/StablehloAttrs.h.inc"
+#endif  // ENUMS_AND_ATTRS
+
+namespace mlir {
+namespace stablehlo {
+
+// Check if the dimension size is dynamic.
```

Comment:
The utils functions are put here temporarily (there's a comment nearby) and will be removed when all shape functions are moved: these utils should not be public.
In `cpp` file, the utils have to be at top of file, because they need to be declared / defined before used. So in this header, I'd prefer put them in the same position. 
Besides, `struct WindowDimension {` can't be moved below as it's used in the signature of shape functions.
Please let me know if you have other thoughts.

Reply:
"So in this header, I'd prefer put them in the same position". Sounds good.
"Besides, `struct WindowDimension` can't be moved below as it's used in the signature of shape functions". How do you mean? It's only used in helpers, right?